lib/gateway-kernel.coffee

```diff
@@ -0,0 +1,94 @@
+child_process = require 'child_process'
+path = require 'path'
+
+_ = require 'lodash'
+uuid = require 'uuid'
+services = require('./jupyter-js-services-shim')
+
+Kernel = require './kernel'
+
+class GatewayKernel extends Kernel
+    constructor: (kernelSpec, session) ->
+        super kernelSpec
+
+        @session = session
+        @session.statusChanged.connect(@_onStatusChange, this)
+        @_onStatusChange() # Set initial status correctly
+
+    interrupt: ->
+        @session.kernel.interrupt()
+
+    restart: ->
+        @session.kernel.restart()
+
+    _onStatusChange: ->
+        @statusView.setStatus @session.status
+
+    _onIOPub: (message, onResults) ->
+        msg_type = message.header.msg_type
+
+        # TODO(nikita): implement support for watches
+
+        if onResults?
+            console.log('got msg')
+            console.log(message)
+            result = @_parseIOMessage(message)
+            if result?
+                onResults result
+
+
+    execute: (code, onResults) ->
+        future = @session.kernel.execute(
+            code: code
+        )
+
+        future.onIOPub = (message) =>
+            @_onIOPub(message, onResults)
+
+        future.onReply = (message) ->
+            if message.content.status is 'error'
+                return
+            result =
+                data: 'ok'
+                type: 'text'
+                stream: 'status'
+            onResults(result)
+
+    executeWatch: (code, onResults) ->
+        # TODO(nikita): implement watches
+        # Also consider switching to user_expressions for watches
+        console.log 'WARNING: Watches are not supported by remote kernels'
+
+    complete: (code, onResults) ->
+        @session.kernel.complete(
+            code: code
+            cursor_pos: code.length
+        ).then((message) ->
+            onResults message.content
+            )
+
+    inspect: (code, cursor_pos, onResults) ->
+        @session.kernel.inspect(
+            code: code
+            cursor_pos: cursor_pos
+            detail_level: 0
+        ).then((message) ->
+            onResults
+                data: message.content.data
+                found: message.content.found
+            )
+
+    addWatchCallback: (watchCallback) ->
+        console.log 'WARNING: Watches are not supported by remote kernels'
+        @watchCallbacks.push(watchCallback)
+
+    destroy: ->
+        super
+        console.log 'destroying jupyter-js-services Session'
+        @session.dispose()
+
+module.exports =
+    gatewayKernelFromSession: (session, callback) ->
```

Comment:
The instance of `Session` is created inside `session-picker.coffee`, but it needs to be converted into a `GatewayKernel` somewhere down the line.

I decided that the `GatewayKernel` constructor should be called from `KernelManager`. But constructors must be sync while `getKernelSpec` is async (and needs to happen before the Kernel is considered fully constructed). Hence this method, which is essentially an async constructor.

If you want to get rid of this method, where do you suggest inlining this code? If it's inlined into `KernelManager`, suddenly there will be lines of code in `KernelManager` that can't be understood without reading the `jupyter-js-services` API docs. I chose to limit `jupyter-js-services` API access to just two files (`session-picker.coffee` and `gateway-kernel.coffee`), instead of spread out all over the place.